File: android_mobile/src/main/java/com/alexstyl/specialdates/upcoming/UpcomingEventsModule.java
package com.alexstyl.specialdates.upcoming;

import android.content.Context;

import com.alexstyl.resources.Colors;
import com.alexstyl.specialdates.CrashAndErrorTracker;
import com.alexstyl.specialdates.Strings;
import com.alexstyl.specialdates.UpcomingEventsJobCreator;
import com.alexstyl.specialdates.analytics.Analytics;
import com.alexstyl.specialdates.date.Date;
import com.alexstyl.specialdates.events.bankholidays.AndroidBankHolidaysPreferences;
import com.alexstyl.specialdates.events.bankholidays.BankHolidayProvider;
import com.alexstyl.specialdates.events.bankholidays.GreekBankHolidaysCalculator;
import com.alexstyl.specialdates.events.namedays.NamedayUserSettings;
import com.alexstyl.specialdates.events.namedays.calendar.OrthodoxEasterCalculator;
import com.alexstyl.specialdates.events.namedays.calendar.resource.NamedayCalendarProvider;
import com.alexstyl.specialdates.events.peopleevents.CompositePeopleEventsProvider;
import com.alexstyl.specialdates.events.peopleevents.PeopleEventsUpdater;
import com.alexstyl.specialdates.events.peopleevents.UpcomingEventsViewRefresher;
import com.alexstyl.specialdates.facebook.FacebookUserSettings;
import com.alexstyl.specialdates.home.HomeNavigator;
import com.alexstyl.specialdates.permissions.MementoPermissions;

import dagger.Module;
import dagger.Provides;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

@Module
public class UpcomingEventsModule {

    @Provides
    UpcomingEventsProvider providesUpcomingEventsProviderWithAds(Context context,
                                                                 Strings strings,
                                                                 Colors colors,
                                                                 NamedayUserSettings namedayUserSettings,
                                                                 NamedayCalendarProvider namedayCalendarProvider,
                                                                 CompositePeopleEventsProvider eventsProvider) {
        Date date = Date.Companion.today();
        return new CompositeUpcomingEventsProvider(
                eventsProvider,
                namedayUserSettings,
                namedayCalendarProvider,
                AndroidBankHolidaysPreferences.newInstance(context),
                new BankHolidayProvider(new GreekBankHolidaysCalculator(OrthodoxEasterCalculator.INSTANCE)),
                new UpcomingEventRowViewModelFactory(
                        date,
                        colors,
                        new AndroidUpcomingDateStringCreator(strings, date, context),
                        new ContactViewModelFactory(colors, strings)
                )
        );
    }

    @Provides
    HomeNavigator navigator(Analytics analytics, Strings strings, CrashAndErrorTracker tracker, FacebookUserSettings facebookUserSettings) {
        return new HomeNavigator(analytics, strings, facebookUserSettings, tracker);
    }

    @Provides
    UpcomingEventsJobCreator jobCreator(PeopleEventsUpdater updater, UpcomingEventsViewRefresher refresher) {
        return new UpcomingEventsJobCreator(updater);
    }

    @Provides
    UpcomingEventsPresenter presenter(MementoPermissions permissionsChecker, UpcomingEventsProvider provider) {
        Date today = Date.Companion.today();
        return new UpcomingEventsPresenter(
                today,
                permissionsChecker,
                provider,
                Schedulers.io(),
                AndroidSchedulers.mainThread()
        );
    }
}

File: android_mobile/src/main/java/com/alexstyl/specialdates/upcoming/UpcomingEventsModule.kt
package com.alexstyl.specialdates.upcoming

import android.content.Context
import com.alexstyl.resources.Colors
import com.alexstyl.specialdates.CrashAndErrorTracker
import com.alexstyl.specialdates.EasyPreferences
import com.alexstyl.specialdates.R
import com.alexstyl.specialdates.Strings
import com.alexstyl.specialdates.analytics.Analytics
import com.alexstyl.specialdates.date.Date
import com.alexstyl.specialdates.events.bankholidays.BankHolidayProvider
import com.alexstyl.specialdates.events.bankholidays.BankHolidaysUserSettings
import com.alexstyl.specialdates.events.bankholidays.GreekBankHolidaysCalculator
import com.alexstyl.specialdates.events.namedays.NamedayUserSettings
import com.alexstyl.specialdates.events.namedays.calendar.resource.NamedayCalendarProvider
import com.alexstyl.specialdates.events.peopleevents.PeopleEventsProvider
import com.alexstyl.specialdates.facebook.FacebookUserSettings
import com.alexstyl.specialdates.home.HomeNavigator
import com.alexstyl.specialdates.permissions.MementoPermissions
import com.alexstyl.specialdates.support.AskForSupport
import com.alexstyl.specialdates.support.CallForRatingPreferences
import com.alexstyl.specialdates.theming.AttributeExtractor
import dagger.Module
import dagger.Provides
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers

@Module
class UpcomingEventsModule {

    @Provides
    fun upcomingEventsProvider(context: Context,
                               strings: Strings,
                               colors: Colors,
                               namedayUserSettings: NamedayUserSettings,
                               namedayCalendarProvider: NamedayCalendarProvider,
                               eventsProvider: PeopleEventsProvider,
                               bankHolidaysUserSettings: BankHolidaysUserSettings,
                               greekBankHolidaysCalculator: GreekBankHolidaysCalculator): UpcomingEventsProvider {
        val date = Date.today()
        return CompositeUpcomingEventsProvider(
                eventsProvider,
                namedayUserSettings,
                namedayCalendarProvider,
                bankHolidaysUserSettings,
                BankHolidayProvider(greekBankHolidaysCalculator),
                UpcomingEventRowViewModelFactory(
                        date,
                        colors,
                        AndroidUpcomingDateStringCreator(strings, date, context),
                        ContactViewModelFactory(colors, strings)
                )
        )
    }

    @Provides
    fun navigator(analytics: Analytics,
                  strings: Strings,
                  tracker: CrashAndErrorTracker,
                  facebookUserSettings: FacebookUserSettings,
                  extractor: AttributeExtractor): HomeNavigator {
        return HomeNavigator(analytics, strings, facebookUserSettings, tracker, extractor)
    }

    @Provides
    fun presenter(permissionsChecker: MementoPermissions, provider: UpcomingEventsProvider): UpcomingEventsPresenter {
        val today = Date.today()
        return UpcomingEventsPresenter(
                today,
                permissionsChecker,
                provider,
                Schedulers.io(),
                AndroidSchedulers.mainThread()
        )
    }

    @Provides
    fun askForSupport(context: Context): AskForSupport {
        val privatePreferences = EasyPreferences.createForPrivatePreferences(context, R.string.pref_call_to_rate)
        val preferences = CallForRatingPreferences(privatePreferences)
        return AskForSupport(preferences)
    }
}
